AEP/aep-00002.md

```diff
@@ -0,0 +1,141 @@
+- Services and Messages should be named appropriately.
+- Enums should have itself as prefix.
+- Only leading comments are considered by Buf, so proto files should include leading comments for everything.
+
+See [Buf Linter Rules](https://buf.build/docs/lint/rules) for an extensive description of each rule used.
+
+## Avoid using `oneof`
+
+gPRC `oneof` means that a value can be one of several messages.
+It highly increase the complexity of using a message when they are used in streams and/or responses.
+
+When used in streams, `oneof` messages are complicated to use since messages could need to be sent in a specific order.
+The usage of this kind of method could be very difficult without appropriate documentation and state machines.
+
+When used in responses, every case should be accounted for even for the nominal case.
+It adds a lot of complexity especially in cases of errors included in `oneof` responses.
+Clients have to manage every case with the possibility of most of the cases being impossible to recover from or being useless.
+
+## Avoid using streams or provide unitary alternative
+
+Some systems does not support streaming such as [gRPC web](https://github.com/grpc/grpc-web) (gRPC web does not support streaming from client to server).
+Therefore we should avoid stream based RPC methods especially in the control plane.
+Stream based methods can still be added if an alternative with unitary calls is also added.
+They can be useful for some optimizations so they are not completely banned.
+
+## Error handling
+
+RPC methods should have an error message that will be sent as response metadata when an error happened during the call.
+This aims to simplify response messages as much as possible for the nominal case.
+
+```proto
+// Exemple response when call is sucessful
+message MyResponse {
+    // The internal data that is sent client-side
+    string my_string_data = 1;
+}
+
+// Exemple error that will be sent to client as response metadata included in the rpc error
+// It may include oneofs when there are several possibilities
+// This should be used in conjonction with gRPC status codes and exceptions
+message MyError {
+    // error that occured
+    string error_details = 1;
+}
+```
+
+## gRPC JSON transcoding
+
+gRPC JSON transcoding is a capability that allows to use JSON REST-like APIs to interact with a service implemented in gRPC.
+It is suggested to make it available for the control plane so that clients can use REST-like request to interact with ArmoniK as it will simplify the use of ArmoniK in environments where gRPC clients are not available or difficult to use.
+
+# Backwards Compatibility
+
+If breaking changes are introduced in new proposal, API version should be increased.
+
+# Reference Implementation
+
+A reference implementation will come as a Pull Request in ArmoniK.Api repository and this will be a first draft of our APIs that follow these guidelines.
```

Comment:
```suggestion
A reference implementation will come as a Pull Request in ArmoniK.Api repository and this will be a first draft of our APIs that follow these guidelines. <!-- markdownlint-disable-line case-police -->
```